// word-frequency-analyzer/bin/src/constants/word_frequency_analyzer/http_api_errors.js
(function() {
  var WORD_FREQUENCY_ANALYZER_HTTP_API_ERRORS;

  WORD_FREQUENCY_ANALYZER_HTTP_API_ERRORS = {
    INCOMING_STREAM_PARSE_ERROR: function(trace) {
      return {
        code: 'IS01',
        message: "Couldn't parse incoming request data stream",
        trace: trace
      };
    },
    PARSER_HEADER_OPTIONS_PARSE_ERROR: function(trace) {
      return {
        code: 'PH01',
        message: "Couldn't parse parser's incoming header options",
        trace: trace
      };
    },
    ANALYZE_DOCUMENT_ERROR: function(trace) {
      return {
        code: 'AD01',
        message: "Was not able to analyze the incoming document data",
        trace: trace
      };
    },
    HTTP_METHOD_ERROR: function(trace) {
      return {
        code: 'HM01',
        message: "Can not execute the given path using the given HTTP method",
        trace: trace
      };
    },
    METHOD_NOT_EXIST: function(trace) {
      return {
        code: 'HM02',
        message: "The method requested to execute does not exist",
        trace: trace
      };
    }
  };

  module.exports = WORD_FREQUENCY_ANALYZER_HTTP_API_ERRORS;

}).call(this);
